File: sdk-ts/src/retry.ts
export const RETRYABLE_STATUS = new Set([429, 502, 503, 504]);

const DEFAULT_MAX_ATTEMPTS = 3;
const DEFAULT_INITIAL_DELAY_MS = 250;
const DEFAULT_MAX_DELAY_MS = 30_000;
const DEFAULT_JITTER_FACTOR = 0.3;

export interface RetryPolicyOptions {
  maxAttempts?: number;
  initialDelayMs?: number;
  maxDelayMs?: number;
  jitterFactor?: number;
}

export class RetryPolicy {
  readonly maxAttempts: number;
  readonly initialDelayMs: number;
  readonly maxDelayMs: number;
  readonly jitterFactor: number;

  constructor(options: RetryPolicyOptions = {}) {
    this.maxAttempts = options.maxAttempts ?? DEFAULT_MAX_ATTEMPTS;
    this.initialDelayMs = options.initialDelayMs ?? DEFAULT_INITIAL_DELAY_MS;
    this.maxDelayMs = options.maxDelayMs ?? DEFAULT_MAX_DELAY_MS;
    this.jitterFactor = options.jitterFactor ?? DEFAULT_JITTER_FACTOR;
  }

  computeDelay(attempt: number, retryAfterMs?: number): number {
    if (retryAfterMs != null) {
      return Math.min(Math.max(retryAfterMs, 0), this.maxDelayMs);
    }
    const base = Math.min(this.initialDelayMs * 2 ** attempt, this.maxDelayMs);
    if (this.jitterFactor === 0) {
      return base;
    }
    const jitterRange = base * this.jitterFactor;
    const jitter = (Math.random() * jitterRange * 2) - jitterRange;
    return Math.max(0, Math.min(base + jitter, this.maxDelayMs));
  }
}

export function isRetryableMethod(
  method: string,
  headers?: HeadersInit,
): boolean {
  const normalizedMethod = method.toUpperCase();
  if (["GET", "HEAD", "PUT", "DELETE", "OPTIONS"].includes(normalizedMethod)) {
    return true;
  }
  if (normalizedMethod !== "POST" || headers == null) {
    return false;
  }
  if (headers instanceof Headers) {
    return headers.has("Idempotency-Key");
  }
  if (Array.isArray(headers)) {
    return headers.some(([key]) => key.toLowerCase() === "idempotency-key");
  }
  return Object.keys(headers).some((key) => key.toLowerCase() === "idempotency-key");
}

File: sdk/agentflow/retry.py
import random
from collections.abc import Mapping, Sequence
from dataclasses import dataclass

RETRYABLE_STATUS = frozenset({429, 502, 503, 504})
DEFAULT_MAX_ATTEMPTS = 3
DEFAULT_INITIAL_DELAY_S = 0.25
DEFAULT_MAX_DELAY_S = 30.0
DEFAULT_JITTER_FACTOR = 0.3


@dataclass(frozen=True)
class RetryPolicy:
    max_attempts: int = DEFAULT_MAX_ATTEMPTS
    initial_delay_s: float = DEFAULT_INITIAL_DELAY_S
    max_delay_s: float = DEFAULT_MAX_DELAY_S
    jitter_factor: float = DEFAULT_JITTER_FACTOR

    def compute_delay(self, attempt: int, retry_after_s: float | None = None) -> float:
        if retry_after_s is not None:
            return min(max(retry_after_s, 0.0), self.max_delay_s)
        base = min(self.initial_delay_s * (2**attempt), self.max_delay_s)
        if self.jitter_factor == 0:
            return base
        jitter_range = base * self.jitter_factor
        jitter = random.uniform(-jitter_range, jitter_range)
        return max(0.0, min(base + jitter, self.max_delay_s))


def is_retryable_method(
    method: str,
    headers: Mapping[str, str] | Sequence[tuple[str, str]] | None = None,
) -> bool:
    normalized_method = method.upper()
    if normalized_method in {"GET", "HEAD", "PUT", "DELETE", "OPTIONS"}:
        return True
    if normalized_method != "POST" or headers is None:
        return False
    if isinstance(headers, Mapping):
        return any(key.lower() == "idempotency-key" for key in headers)
    return any(key.lower() == "idempotency-key" for key, _ in headers)
